common/src/main/cpp/classes/openxr_fb_spatial_anchor_manager.cpp

```diff
@@ -0,0 +1,505 @@
+	} else {
+		p_spatial_entity->connect("openxr_fb_spatial_entity_set_component_enabled_completed", callable_mp(this, &OpenXRFbSpatialAnchorManager::_on_anchor_untrack_enable_storable_completed).bind(p_spatial_entity), CONNECT_ONE_SHOT);
+		p_spatial_entity->set_component_enabled(OpenXRFbSpatialEntity::COMPONENT_TYPE_STORABLE, true);
+	}
+}
+
+void OpenXRFbSpatialAnchorManager::_on_anchor_untrack_enable_storable_completed(bool p_succeeded, OpenXRFbSpatialEntity::ComponentType p_component, bool p_enabled, const Ref<OpenXRFbSpatialEntity> &p_spatial_entity) {
+	if (!p_succeeded) {
+		// If we couldn't make it storable, just exit silently since we were trying to remove it anyway.
+		return;
+	}
+
+	p_spatial_entity->connect("openxr_fb_spatial_entity_erased", callable_mp(this, &OpenXRFbSpatialAnchorManager::_on_anchor_erase_completed).bind(p_spatial_entity), CONNECT_ONE_SHOT);
+	p_spatial_entity->erase_from_storage(OpenXRFbSpatialEntity::STORAGE_LOCAL);
+}
+
+void OpenXRFbSpatialAnchorManager::_on_anchor_erase_completed(bool p_succeeded, OpenXRFbSpatialEntity::StorageLocation p_location, const Ref<OpenXRFbSpatialEntity> &p_spatial_entity) {
+	ERR_FAIL_COND_MSG(!p_succeeded, vformat("Unable to erase spatial anchor %s.", p_spatial_entity->get_uuid()));
+}
+
+Error OpenXRFbSpatialAnchorManager::save_anchors_to_local_file() {
+	Ref<FileAccess> file = FileAccess::open(local_file_path, FileAccess::WRITE);
+	if (file.is_null()) {
+		return FileAccess::get_open_error();
+	}
+
+	Dictionary anchor_data;
+	for (const KeyValue<StringName, Anchor> &E : anchors) {
+		anchor_data[E.key] = E.value.entity->get_custom_data();
+	}
+
+	file->store_string(JSON::stringify(anchor_data));
+
+	return OK;
+}
+
+Error OpenXRFbSpatialAnchorManager::load_anchors_from_local_file() {
+	ERR_FAIL_COND_V(!xr_origin, FAILED);
+	ERR_FAIL_COND_V(anchors.size() > 0, FAILED);
+
+	Ref<FileAccess> file = FileAccess::open(local_file_path, FileAccess::READ);
+	if (file.is_null()) {
+		return FileAccess::get_open_error();
+	}
+
+	Ref<JSON> json;
+	json.instantiate();
+	Error parse_error = json->parse(file->get_as_text());
+	if (parse_error != OK) {
+		return parse_error;
+	}
+
+	Dictionary anchor_data = json->get_data();
+
+	Ref<OpenXRFbSpatialEntityQuery> query;
+	query.instantiate();
+	if (erase_unknown_anchors_on_load) {
+		// If we want to clear out unknown anchors, then we need to query everything.
+		query->query_all();
+		query->set_max_results(1000);
+	} else {
+		// Otherwise, we just query the specific anchors we know about.
+		query->query_by_uuid(anchor_data.keys(), OpenXRFbSpatialEntity::STORAGE_LOCAL);
+		query->set_max_results(anchor_data.size());
+	}
+	query->connect("openxr_fb_spatial_entity_query_completed", callable_mp(this, &OpenXRFbSpatialAnchorManager::_on_anchor_load_query_completed).bind(anchor_data, OpenXRFbSpatialEntity::STORAGE_LOCAL, false));
+	query->execute();
+
+	return OK;
+}
```

Comment:
I don't think saving to / loading from a file are functionality that belongs to this class / plugin for several reasons:
- File access differs widely based on the platform. On desktop OSes (e.g: Windows, Linux) it has few limitations whereas on mobile platforms like Android only certain paths are available to write to (`res://` cannot be written to, `user://` can, and only certain `filesystem` paths are available). By providing the functionality in this plugin, we're giving the user the illusion we're smoothing all the platform specific underlying issues, which this PR definitely doesn't, but attempting to do so would deserve its own plugin given the complexity and edge cases involved.
- IO bound operations are typically best handled on a separate thread to avoid blocking the main thread. For a sample like this plugin's demo not concerned with performance, we can perform IO within the same thread, but this is not an approach I'd advocate for more complex apps / games.
- The proposed implementation uses `json` to parse the data; this is fine for a sample like this plugin's demo, but again this is not an approach I'd advocate for more complex apps / games, due to `json` own limitations.

Reply:
So, are you advocating for leaving this problem entirely up to the developer?

My thinking was that we provide a default in order to reduce the boilerplate for getting started, but it is entirely optional. The developer can turn this off by disabling the `persist_in_local_file` property, and then do their own thing.

Anyway, I can certainly just move all this code to the demo. But then when developers want to add this to their own project, they'll either have to copy that same code, or write something new from scratch, making it more work to get started.

Reply:
Yeah, I advocate leaving this up to the developer. Here I use the same approach we use for the Godot proposals; i.e: is this something that the developer can resolve using scripting or is this something that needs to be incorporate into `core`, where `core` here stands for the vendors plugin.

And this definitely falls into the first category, i.e: this is something developers can resolve via scripting. We can provide guidance via the demo, or by putting a gist illustrating this capability.
My expectation is that this would be useful for folks testing the capability, but not when building a full app / game, since in that scenario you'd want the data layer for the app / game to be consolidated (e.g: common approach for storing anchors, user data, game state, etc...).

Reply:
Alright, that makes sense!

In my next update, I'll move the file bits into the demo.